<a href="https://colab.research.google.com/github/smpark0520/ESAA/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%99%84%EB%B2%BD_%EA%B0%80%EC%9D%B4%EB%93%9C_p_39~84.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 04 데이터 핸들링 - 판다스


*   데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리
*   넘파이보다 훨씬 유연하고 편리하게 데이터 핸들링 가능



# 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


- /content/drive/MyDrive 디렉터리에 있는 titanic train.csv 파일 업로드
- read_csv를 입력해 데이터를 DataFrame으로 로딩

In [ ]:
titanic_df=pd.read_csv("/content/drive/MyDrive/titanic train.csv")
print('titanic 변수 type:',type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

- head(3) : 맨 앞 3개의 row를 반환 (기본값은 5개)

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

- shape 변수를 이용해 DataFrame의 행과 열 크기 알 수 있음

In [ ]:
print('DataFrame 크기:',titanic_df.shape)

DataFrame 크기: (891, 12)


- 891개의 row와 12개의 col으로 구성

In [ ]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


- RangeIndex: 891 entries, 0 to 890 -> 전체 row 수를 알 수 있음. 전체 데이터는 891개 rowd임
- Dtype -> 데이터 타입
- 714 non-null -> age칼럼의 891개의 데이터 중 714개가 null이 아니며 177개는 null임
- 2개 칼럼이 float64, 5개 칼럼이 int64, 5개 칼럼이 object 타입임

In [ ]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- describe() 메서드를 입력해 숫자형 칼럼에 대한 개략적인 데이터 분포도 확인 가능

- value_counts() : 해당 칼럼갑스이 유형, 건수 확인
- 지정된 칼럼의 데이터값 건수를 반환

In [ ]:
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


- Pclss값 3이 491개, 1이 216개, 2가 184개

- DataFrame의 [ ] 연산자 내부에 칼럼명을 입력하면해당 칼럼에 해당하는 Series 객체를 반환

In [ ]:
titanic_pclass=titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [ ]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

- 오른쪽은 Series의 해당 칼럼의 데이터값

- value_counts() 를 titanic_df의 Pclass 칼럼만을 값으로 가지는 Series 객체에서 호출하여 메서드의 반환값을 다시 확인

In [ ]:
value_counts=titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


- 왼쪽은 value_counts() 반환 Series의 index
- 오른쪽은 value_counts() 반환 Series의 데이터값

In [ ]:
print('titanic_df 데이터 건수:',titanic_df.shape[0])
print('기본 설정인 dropna=True로 value_counts()')
#value_counts()는 디폴트로 dropna=True이므로 value_counts(dropna=True)와 동일
print(titanic_df['Embarked'].value_counts())
print(titanic_df['Embarked'].value_counts(dropna=False))

titanic_df 데이터 건수: 891
기본 설정인 dropna=True로 value_counts()
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


- 타이타닉 데이터 세트의 Embarked 칼럼의 데이터 값 분포를 value_counts()로 확인
- Embarked 칼럼은 전체 891건 중에 2건의 데이터가 Null
- 'S’, ‘C’, ‘Q’ 값으로 각각 644, 168,77을 반환
- dropna=False : Null 값을 포함하여 value_counts()를 적용

# DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환

넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기

In [ ]:
import numpy as np

col_name1=['col1']
list1=[1,2,3]
array1=np.array(list1)
print('array1.shape:',array1.shape)
#리스트를 이용해 DataFrame 생성
df_list1=pd.DataFrame(list1,columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n',df_list1)
#넘파이 ndarray를 이용해 DataFrame 생성
df_array1=pd.DataFrame(array1,columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n',df_array1)

array1.shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


- 1차원 형태의 리스트와 넘파이 ndarray부터 DataFrame
으로 변환

In [ ]:
#3개의 칼럼명이 필요함
col_name2=['col1','col2','col3']

#2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2=[[1,2,3],
       [11,12,13]]
array2=np.array(list2)
print('array2 shape:',array2.shape )
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame：\n',df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame：\n',df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame：
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame：
    col1  col2  col3
0     1     2     3
1    11    12    13


-  2차원 형태의 데이터를 기반으로 DataFrame을 생성
- 2행 3열 형태의 리스트와 ndarray를 기반으로 DataFrame을 생성하므로 칼럼명은 3개가 필요

In [ ]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형（또는 ndarray） 칼럼 데이터로 매핑
dict = {'col1':[1, 11], 'col2':[2, 22], 'col3':[3, 33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame：\n',df_dict)

딕셔너리로 만든 DataFrame：
    col1  col2  col3
0     1     2     3
1    11    22    33


- 딕셔너리를 DataFrame으로 변환
- Key 값은 칼럼명, 각 Value는 각 칼럼 데이터로 매핑



DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

In [ ]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입:',type(array3), 'df_dict.values shape：',array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape： (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [ ]:
# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입:',type(list3))
print(list3)
# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


# DataFrame의 칼럼 데이터 세트 생성과 수정

- Titanic DataFrame의 새로운 칼럼 Age_0을 추가하고 일괄적으로 0 값을 할당

In [ ]:
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

- 기존 칼럼 Series를 가공해 새로운 칼럼 Series인 Age_by_10과 Family_No가 새롭게 DataFrame에 추가

In [ ]:
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

-  Age_by_10' 칼럼 값을 일괄적으로 기존 값+100으로 업데이트

In [ ]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

# DataFrame 데이터 삭제

- axis 0은 로우 방향 축, 특정 로우를 드롭
- axis 1은 칼럼 방향 축 , 특정 칼럼을 드롭

- 'Age_0' 칼럼을 삭제

In [ ]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

- 원본 Titanic DataFrame을 다시 확인했더니 삭제됐다고 생각했던 ‘Age_0’ 칼럼이 여전히 존재
- Inplace=False이면 자기 자신의 DataFrame의 데이터는 삭제하지 않으며, 삭제된 결과 DataFrame을 반환

- inplace=True로 설정하면 자신의 DataFrame의 데이터를 삭제

In [ ]:
drop_result=titanic_df.drop(['Age_0','Age_by_10','Family_No'], axis=1,inplace=True)
print('inplace=True 로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True 로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

- ‘Age_0’, ‘Age_by_10’, ‘Family_No’ 세 개의 칼럼이 모두 삭제

In [ ]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth',15)
print( '#### before axis 0 drop ####')
print(titanic_df.head(3))
titanic_df.drop([0,1,2], axis=0, inplace=True)
print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


- 이번에는 axis=0으로 설정해 index 0, 1, 2(맨 앞 3개 데이터) 로우를 삭제

-  axis： DataFrame의 로우를 삭제할 때는 axis=0, 칼럼을 삭제할 때는 axis=l으로 설정
-  원본 DataFrame은 유지하고 드롭된 DataFrame을 새롭게 객체 변수로 받고 싶다면 inplace三False로 설정(디폴트 값이 False임)
예: titanic_drop_df = titanic_df.drop('Age_0', axis=1, inplace=False )
-  원본 DataFrame에 드롭된 결과를 적용할 경우에는 inplace=True를 적용.
예: titanic_df.drop('Age_0'axis=1inplace=True )
-  원본 DataFrame에서 드롭된 DataFrame을 다시 원본 DataFrame 객체 변수로 할당하면 원본 DataFrame에서 드롭된 결과를 적용할 경우와 같음(단, 기존 원본 DataFrame 객체 변수는 메모리에서 추후 제거됨).
예: titanic_df = titanic_df.drop(' Age_0', axis=1, inplace=False )


# Index 객체

In [ ]:
# 원본 파일 다시 로딩
titanic_df = pd.read_csv('/content/drive/MyDrive/titanic train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 arrray로 변환
print('Index 객체 array값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

In [ ]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


In [ ]:
indexes[0] = 5

TypeError: Index does not support mutable operations

- 한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없음

In [ ]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series：',sum(series_fair))
print('Fair Series + 3:\n', (series_fair + 3).head(3) )

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series： 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


- Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외

-  reset_index() :새롭게 인덱스를 연속 숫자 형으로 할당
- 기존 인덱스는 ‘index’라는 새로운 칼럼 명으로 추가

In [ ]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [ ]:
print( '### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:',type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print( '### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   Pclass  count
0       3    491
1       1    216
2       2    184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


- drop=Tme로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)

# 데이터 셀렉션 및 필터링

DataFrame의 [ ] 연산자


In [ ]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[ ['Survived', 'Pclass'] ].head(3))
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n',titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

- DataFrame 뒤의 [ ]에는 칼럼명을 지정해야 하는데, 0이 칼럼명이 아니기 때문에 오류 발생

- 처음 2개 데이터를 추출

In [ ]:
titanic_df[ 0:2 ]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

- Pclass 칼럼 값이 3인 데이터 3개만 추출

In [ ]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

DataFrame iloc[ ] 연산자


- 은 인덱스 값으로 숫자값이 아닌 문자열로 ‘one’, ‘two’, ‘three’, ‘four’ 를 가지고 있으며, Name,  Year,
Gender 3개의 칼럼을 가짐

In [ ]:
data = {'Name': ['Chulmin', 'Eunkyung','Jinwoong', 'Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
}
data_df = pd.DataFrame(data, index=['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

-  data_df의 첫 번째 행, 첫 번째 열의 데이터 추출

In [ ]:
data_df.iloc[0,0]

'Chulmin'

-  iloc[ ]의 열 위치에 위치 정숫값이 아닌 칼럼 명칭을 입력하여 오류가 발생함

In [ ]:
# 아래 코드는 오류를 발생시킵니다.
data_df.iloc[0, 'Name']

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

- iloc[ ]의 행 위치에 DataFrame의 인덱스 값인 ‘one’을 입력하여 오류가 발생

In [ ]:
# 아래 코드는 오류를 발생합니다.
data_df.iloc['one', 0]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

-  iloc[：, -1] : 맨 마지
막 칼럼의 값, 즉 타깃 값을 가져옴
- iloc[:, ：-1] : 처음부터 맨 마지막 칼럼을 제외한 모든 칼럼의 값 가져옴

In [ ]:
print("\n 맨 마지막 칼럼 데이터 [:, -1 ] \n", data_df.iloc[:, -1])
print("\n 맨 마지막 칼럼을 제외한 모든 데이터 [:, ：-1] \n", data_df.iloc[: , :-1])


 맨 마지막 칼럼 데이터 [:, -1 ] 
 one        Male
two      Female
three      Male
four       Male
Name: Gender, dtype: object

 맨 마지막 칼럼을 제외한 모든 데이터 [:, ：-1] 
            Name  Year
one     Chulmin  2011
two    Eunkyung  2016
three  Jinwoong  2015
four    Soobeom  2015


DataFrame loc[] 연산자


- 인덱스값이 ‘one’인 행의 칼럼명이 'Name'인 데이터를 추출

In [ ]:
data_df.loc['one', 'Name']

'Chulmin'

In [ ]:
# 다음 코드는 오류를 발생시킵니다.
data_df.loc[0, 'Name']

KeyError: 0

- 첫번째 행 위치의 인덱스 값을 0으로 착각하고 행 위치에 0을 입력했지만 data_df는 0을 인덱스 값으로 가지고 있지 않기에 오류를 발생

In [ ]:
print('위치기반 iloc slicing\n', data_df.iloc[0:1, 0],'\n')
print('명칭기반 loc slicing\n', data_df.loc['one':'two','Name'])

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


- data_df.loc['one':'two','Name'] : 인덱스 값 one부터 two까지 행의 Name칼럼 반환

불린 인덱싱

- 승객 중 나이(Age)가 60세 이상인 데이터를 추출

In [ ]:
titanic_boolean = titanic_df[titanic_df['Age'] >60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

- 60세 이상인 승객의 나이와 이름만 3개 추출

In [ ]:
titanic_df[titanic_df['Age'] > 60][['Name','Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [ ]:
titanic_df.loc[titanic_df['Age'] > 60, ['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

- 나이가 60세 이상이고, 선실 등급이 1등급이며, 성별이 여성인 승객을 추출

In [ ]:
titanic_df[ (titanic_df['Age'] > 60) & (titanic_df['Pclass']==1) &
(titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

In [ ]:
condl = titanic_df['Age'] > 60
cond2 = titanic_df [' Pclass' ]=1
cond3 = titanic_df[ 'Sex' ]=='female'
titanic_df[ condl & cond2 & cond3]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked   Pclass
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S        1
483          484         1       3  Turkula, Mr...  female  63.0      0      0    4134   9.5875   NaN        S        1
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN        1

# 정렬, Aggregation 함수, GroupBy 적용

DataFrame, Series의 정렬 - sort_values()

-  titanic_df를 Name 칼럼으로 오름차순 정렬

In [ ]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked   Pclass
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S        1
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S        1
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S        1

-  Pclass와 Name를 내림차순으로 정렬

In [ ]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'],ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked   Pclass
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S        1
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S        1
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S        1

Aggregation 함수 적용

In [ ]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
 Pclass        891
dtype: int64

In [ ]:
titanic_df[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

groupby() 적용

- groupby(by='Pclass') : Pclass 칼럼 기준으로 GroupBy된 객체를 반환

In [ ]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [ ]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked   Pclass
Pclass                                                                                             
1               216       216   216  216  186    216    216     216   216    176       214      216
2               184       184   184  184  173    184    184     184   184     16       184      184
3               491       491   491  491  355    491    491     491   491     12       491      491

- titanic_df.groupby('Pclass')로 반환된 DataFrameGroupBy 객체에 [['Passengerld','Survived']]로 필터링해 Passengerld와 Survived칼럼에만 count()를 수행

In [ ]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [ ]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

<ipython-input-65-b719d79acb8d>:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])
<ipython-input-65-b719d79acb8d>:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])


max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

- agg() 내에 입력값으로 딕셔너리 형태로
aggregation이 적용될 칼럼들과 aggregation 함수를 입력

In [ ]:
agg_format={'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

# 결손 데이터 처리하기


isna()로 결손 데이터여부 확인

- 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려줌

In [ ]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked   Pclass
0        False     False   False  False  False  False  False  False   False  False   True     False    False
1        False     False   False  False  False  False  False  False   False  False  False     False    False
2        False     False   False  False  False  False  False  False   False  False   True     False    False

- sum()을 호출 시 True는 내부적으로 숫자 1로, False는 숫자 0으로 변환되므로 결손 데이터의 개수를 구할 수 있음

In [ ]:
titanic_df.isna( ).sum( )

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
 Pclass          0
dtype: int64

fillna()로 결손 데이터 대체하기


-  ‘Cabin’ 칼럼의 NaN 값을 ‘C000’으로 대체

In [ ]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked   Pclass
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S        1
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C        1
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S        1

- ‘Age’ 칼럼의 NaN 값을 평균 나이로, 'Embarked’ 칼럼의 NaN 값을 ‘으로 대체해 모든 결손 데이터를 처리

In [ ]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'] .mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna(). sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
 Pclass        0
dtype: int64

# apply lambda 식으로 데이터 가공

In [ ]:
def get_square(a):
    return a**2
print('3의 제곱은:', get_square(3))

3의 제곱은: 9


In [ ]:
lambda_square = lambda x : x ** 2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


- ‘:’의 왼쪽에 있는
x는 입력 인자를 가리키며
- 오른쪽은 입력 인자의 계산식입니다. 오른쪽의 계산식은 결국 반환 값을 의미

In [ ]:
a=[1, 2, 3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

- lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우 map() 함수를 결합

-  ‘Name’ 칼럼의 문자열 개수를 별도의 칼럼인 ‘Name_len’에 생성

In [ ]:
titanic_df['Name_len']= titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

- 나이가 15세 미만이면 'Child', 그렇지 않으면 ‘Adult'로 구분하는 새로운 칼럼 ‘Child_Adult'를 apply lambda를 이용해 생성

In [ ]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <=15 else 'Adult' )
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

- 나이가 15세 이하이면 Child, 15〜60세 사이는 Adult, 61세 이상은 Elderly로 분류

In [ ]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<=15 else ('Adult' if x <= 60
else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Age_cat
Adult      786
Child       83
Elderly     22
Name: count, dtype: int64

- 5살 이하는 Baby, 12살 이하는 Child, 18살 이하는 Teenage, 25살 이하는 Student, 35살 이하는 Young Adult, 60세 이하는 Adult, 그리고 그 이상은 Elderly로 분류

In [ ]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성.
def get_category(age):
    cat = ''
    if age <= 5: cat = 'Baby'
    elif age <= 12: cat = 'Child'
    elif age <= 18: cat = 'Teenager'
    elif age <= 25: cat = 'Student'
    elif age <= 35: cat = 'Young Adult'
    elif age <= 60: cat = 'Adult'
    else : cat = 'Elderly'

    return cat

# lambda 식에 위에서 생성한 get_category( ) 함수를 반환값으로 지정.
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult